### PR TITLE
[GSoC] robust function for affine transform estimation

### PR DESCRIPTION
**Pull request by vpisarev**
### Former  #6615

implements `estimateAffine2D`. estimates affine transformation using robust RANSAC method.

* uses RANSAC framework in calib3d
* includes accuracy test
* uses SVD decomposition for solving 3 point equation

things to consider:
* is solving through SVD fast enough? should we make it symmetric and use EIGEN val decomposition?
* is checking for coplanar points robust enough? (it seems so)

I plan to write perf test to see if make sense to do some tricks and solve with EIGEN val decomp. Checking for coplanar points seems to work well for me. I adapted the check from estimateAffine3D, estimateRigidTransform have some addition checks (checking if points are not too close), but I'm not sure if it is worth to implement those.

I will implement also `estimateAffinePartial2D` that will estimate transformation using only rotation, translation an scaling (similar to `estimateRigidTransform`). I'm only not sure about the name `estimateAffinePartial2D`, any ideas welcome.

cc: @prclibo

### CODE REVIEW

File: modules/calib3d/src/ptsetreg.cpp [hunk 1]
```diff
@@ -562,7 +562,7 @@ class Affine2DEstimatorCallback : public PointSetRegistrator::Callback
             double a = F[0]*f.x + F[1]*f.y + F[2] - t.x;
             double b = F[3]*f.x + F[4]*f.y + F[5] - t.y;
 
-            errptr[i] = (float)std::sqrt(a*a + b*b);
+            errptr[i] = (float)(a*a + b*b);
```
Comment (prclibo) [inline modules/calib3d/src/ptsetreg.cpp]: Why is this changed?

Comment (hrnr) [inline modules/calib3d/src/ptsetreg.cpp]: That is the same change as in #6768 to stay consistent when comparing thresholds.

**Comment by mself** [comment]
Thanks for adding robust estimation to the 2D models.  I always wondered why there was RANSAC support for the 3D models, but not the 2D ones.

Have you considered adding support for weighted least squares?  The feature tracking functions return a match strength, and that could be used to weight the matches so that blurry or moving points aren't weighted as strongly as sharper matches.  I think that could also help improve robustness.

**Comment by hrnr** [comment]
I'm not thinking about adding weighted least squares (no place to accept weights in current API), but I'm thinking about supporting LMEDS (least median) method and adding Levenberg–Marquardt final step to fine-tune the transformation. I'd like the new fuctions to be similar to `findHomography` and `estimateAffine3D`.

I have never used *weighted* least squares, but aren't the weighted least squares trivial to implement? Like multiply both sides of equation with weights and call `cv::solve` as always? Or is it more tricky in practise?

**Comment by mself** [comment]
Yes, I think weighted least squares is as straightforward as you suggest.  In findHomography, I recall that there is a post-least-squares "optimization" step that I didn't understand (maybe that's the Levenberg–Marquardt step you mentioned).  I wasn't sure how that would need to be modified to account for the weights or else I would have done this myself.

**Comment by hrnr** [comment]
Incorporating weights into RANSAC and Levenberg–Marquardt makes sense to me. For that current framework in calib3d is ok. 

For optimization step you mean [this](https://github.com/opencv/opencv/blob/master/modules/calib3d/src/fundam.cpp#L404)? That's LevMarq alg. For that you need to implement a `LMSolver::Callback` which computes Jacobian and errors (in both you would need to incorporate weights).

It would be quite similar to what I'm doing for affine estimation functions.

**Comment by hrnr** [comment]
replaced by #6933
